keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,544 @@
+    SetReplicas(0).
+    SetTemplate(
+      &v1apply.PodTemplate().
+        SetSpec(&v1apply.SetPodSpec().
+          SetContainers(v1apply.ContainerList{
+            v1apply.Container().
+              SetName("nginx").
+              SetImage("nginx:1.14.2")
+            v1apply.Container().
+              SetName("sidecar").
+          })
+        )
+      )
+    )
+  )
+```
+
+#### Comparison of alternatives
+
+See https://github.com/kubernetes/kubernetes/pull/95988 for a working implementation
+of alterative 1 and https://github.com/jpbetz/kubernetes/tree/apply-client-go-builders
+for a working implementation of alternative 2.
+
+Of the two leading alternatives--"builders" and "structs with pointers"--we implemented
+prototypes of both. They had roughly equivalent performance, and no differences
+in their capabilities. The choice between the two is primarily one of asthetics/ergonomics.
+
+Some of the feedback we have heard from the community:
+
+- "structs with pointers" feels more go idiomatic and more closely aligned with
+  the go structs used for Kubernetes types both for builtin types and by
+  Kubebuilder.
+- It's nice how "builders" are clearly visually distinct from the main go types.
+- Having to use a utility library to wrap literal values as pointers for the
+  "structs with pointers" is not a big deal. I'm already familiar
+  with having to do this in go and once I learn use a utility library for it
+  I'm all set.
+- The "builders" are awkward to use in an IDE. I felt like I was fighting with
+  my IDE to get chain function calls and organize them hierarchally as expected
+  by this approach.
+
+TODO: We are providing the developer community with a fork of controller-tools
+that will allow them to generate apply configuration types that have both the
+alternatives. Our goal is to get feedback from developers after they try out the
+generated apply configurations and use that to inform our decision.
+
+While it is possible to generate types that have both the pointer fields exposed
+and the builder functions, we would prefer to provide clear guidance to the
+community on how apply configurations should be represented in go and encourage
+consistent use of only one of these approaches.
+
+#### DeepCopy support
+
+If "structs with pointers" approach is used, the existing deepcopy-gen
+can be used to generate deep copy impelemntations for the generated apply
+configuration types.
+
+#### Code Generator Changes
+
+hack/update-codegen.sh and hack/verify-codegen.sh will be updated to generate
+the apply functions and apply configuration types.
+
+##### Addition of applyconfiguration-gen
+
+- Add staging/src/k8s.io/code-generator/cmd/applyconfigurations-gen
+- Generates into staging/vendor/k8s.io/client-go/applyconfigurations/
+- Only generate builders for struct types reachable from the types that have the +clientgen annotation
+- Don't generate builders for MarshalJSON types (Quantity, IntOrString)
```

Comment:
People construct these directly?

Reply:
You're right, that's a bad example (although they do implement MarshalJSON). Looks like Duration and Time are better examples, I'll use them here.